src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -16,34 +16,75 @@ internal sealed class SourceGeneratedDocumentState : DocumentState
         public ISourceGenerator SourceGenerator { get; }
 
         public static SourceGeneratedDocumentState Create(
-            GeneratedSourceResult generatedSourceResult,
+            SourceGeneratedDocumentState? previousState,
+            string hintName,
+            SourceText generatedSourceText,
+            SyntaxTree generatedSyntaxTree,
             DocumentId documentId,
             ISourceGenerator sourceGenerator,
             HostLanguageServices languageServices,
-            SolutionServices solutionServices)
+            SolutionServices solutionServices,
+            CancellationToken cancellationToken)
         {
-            var textAndVersion = TextAndVersion.Create(generatedSourceResult.SourceText, VersionStamp.Create());
-            ValueSource<TextAndVersion> textSource = new ConstantValueSource<TextAndVersion>(textAndVersion);
+            var tree = generatedSyntaxTree;
+            var options = tree.Options;
+            var filePath = tree.FilePath;
+
+            bool sourceChanged;
+            ValueSource<TextAndVersion> textSource;
 
-            var tree = generatedSourceResult.SyntaxTree;
+            // We can reuse the previous TextAndVersion if the checksum from the text matches.
+            if (previousState is null
+                || !previousState.TryGetTextAndVersion(out var textAndVersion)
+                || Checksum.From(textAndVersion.Text.GetChecksum()) != Checksum.From(generatedSourceText.GetChecksum()))
```

Comment:
nit: cnosider just a function that compares the results of GetChecksum so you don't have to wrap in another Checksum  object.